File: backend/src/main/java/br/com/calciolari/datahub/analytics/application/DashboardQueryService.java
package br.com.calciolari.datahub.analytics.application;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.springframework.stereotype.Service;

import br.com.calciolari.datahub.analytics.api.DashboardDtos.DailyPoint;
import br.com.calciolari.datahub.analytics.api.DashboardDtos.DashboardResponse;
import br.com.calciolari.datahub.sales.infrastructure.persistence.SaleRepository;

@Service
public class DashboardQueryService {

	private final SaleRepository saleRepository;

	public DashboardQueryService(SaleRepository saleRepository) {
		this.saleRepository = saleRepository;
	}

	public DashboardResponse summarize(UUID productId, LocalDateTime from, LocalDateTime to) {
		LocalDateTime fromBound = from == null ? LocalDateTime.of(1, 1, 1, 0, 0) : from;
		LocalDateTime toBound = to == null ? LocalDateTime.of(9999, 12, 31, 23, 59, 59) : to;
		BigDecimal revenue = nullToZero(saleRepository.sumPublishedRevenue(productId, fromBound, toBound));
		BigDecimal quantity = nullToZero(saleRepository.sumPublishedQuantity(productId, fromBound, toBound));
		long salesCount = saleRepository.countPublishedSales(productId, fromBound, toBound);
		long itemCount = saleRepository.countPublishedItems(productId, fromBound, toBound);
		String averageTicket = salesCount > 0
				? revenue.divide(BigDecimal.valueOf(salesCount), 2, RoundingMode.HALF_UP).toPlainString()
				: null;

		Map<LocalDate, BigDecimal[]> byDay = new LinkedHashMap<>();
		for (Object[] row : saleRepository.publishedMovementFacts(productId, fromBound, toBound)) {
			LocalDateTime occurredAt = (LocalDateTime) row[0];
			BigDecimal qty = nullToZero((BigDecimal) row[1]);
			BigDecimal tot = nullToZero((BigDecimal) row[2]);
			LocalDate day = occurredAt.toLocalDate();
			byDay.compute(day, (d, acc) -> {
				if (acc == null) {
					return new BigDecimal[] {qty, tot};
				}
				acc[0] = acc[0].add(qty);
				acc[1] = acc[1].add(tot);
				return acc;
			});
		}
		List<DailyPoint> daily = new ArrayList<>();
		byDay.entrySet().stream()
				.sorted(Map.Entry.comparingByKey())
				.forEach(entry -> daily.add(new DailyPoint(
						entry.getKey().toString(),
						entry.getValue()[0].stripTrailingZeros().toPlainString(),
						entry.getValue()[1].stripTrailingZeros().toPlainString())));

		return new DashboardResponse(
				decimal(revenue),
				decimal(quantity),
				salesCount,
				itemCount,
				averageTicket,
				iso(saleRepository.minPublishedOccurredAt(productId, fromBound, toBound)),
				iso(saleRepository.maxPublishedOccurredAt(productId, fromBound, toBound)),
				daily);
	}

	private static BigDecimal nullToZero(BigDecimal value) {
		return value == null ? BigDecimal.ZERO : value;
	}

	private static String decimal(BigDecimal value) {
		return value == null ? null : value.stripTrailingZeros().toPlainString();
	}

	private static String iso(LocalDateTime value) {
		return value == null ? null : value.toString();
	}
}

File: backend/src/Calciolari.DataHub/Analytics/Application/DashboardQueryService.cs
using Calciolari.DataHub.Persistence;
using Calciolari.DataHub.Shared.Api;
using Microsoft.EntityFrameworkCore;

namespace Calciolari.DataHub.Analytics.Application;

public sealed class DashboardQueryService
{
    private readonly DataHubDbContext _db;

    public DashboardQueryService(DataHubDbContext db)
    {
        _db = db;
    }

    public DashboardResponse Summarize(Guid? productId, DateTime? from, DateTime? to)
    {
        var fromBound = DateTime.SpecifyKind(
            from ?? new DateTime(1, 1, 1, 0, 0, 0, DateTimeKind.Unspecified), DateTimeKind.Unspecified);
        var toBound = DateTime.SpecifyKind(
            to ?? new DateTime(9999, 12, 31, 23, 59, 59, DateTimeKind.Unspecified), DateTimeKind.Unspecified);

        var items = PublishedItems(productId, fromBound, toBound).ToList();
        var revenue = items.Sum(i => i.Total);
        var quantity = items.Sum(i => i.Quantity);
        var salesCount = items.Select(i => i.SaleId).Distinct().Count();
        var itemCount = items.Count;
        var averageTicket = salesCount > 0
            ? decimal.Round(revenue / salesCount, 2, MidpointRounding.AwayFromZero).ToString(System.Globalization.CultureInfo.InvariantCulture)
            : null;

        var byDay = new SortedDictionary<DateOnly, (decimal Qty, decimal Tot)>();
        foreach (var row in items.Where(i => i.OccurredAt is not null))
        {
            var day = DateOnly.FromDateTime(row.OccurredAt!.Value);
            if (byDay.TryGetValue(day, out var acc))
            {
                byDay[day] = (acc.Qty + row.Quantity, acc.Tot + row.Total);
            }
            else
            {
                byDay[day] = (row.Quantity, row.Total);
            }
        }

        var daily = byDay.Select(entry => new DailyPoint(
            entry.Key.ToString("yyyy-MM-dd"),
            DecimalText.StripTrailingZerosToPlain(entry.Value.Qty),
            DecimalText.StripTrailingZerosToPlain(entry.Value.Tot))).ToList();

        var topProducts = items
            .GroupBy(i => i.ProductId)
            .Select(g => new
            {
                ProductId = g.Key,
                Quantity = g.Sum(x => x.Quantity),
                Revenue = g.Sum(x => x.Total)
            })
            .OrderByDescending(g => g.Revenue)
            .Take(5)
            .ToList()
            .Select(g =>
            {
                var product = _db.Products.AsNoTracking().Single(p => p.Id == g.ProductId);
                return new TopProduct(
                    product.Id,
                    product.Name,
                    product.ExternalId,
                    DecimalText.StripTrailingZerosToPlain(g.Quantity),
                    DecimalText.StripTrailingZerosToPlain(g.Revenue));
            })
            .ToList();

        DateTime? first = items.Where(i => i.OccurredAt is not null).Select(i => i.OccurredAt).Min();
        DateTime? last = items.Where(i => i.OccurredAt is not null).Select(i => i.OccurredAt).Max();

        return new DashboardResponse(
            DecimalText.StripTrailingZerosToPlain(revenue),
            DecimalText.StripTrailingZerosToPlain(quantity),
            salesCount,
            itemCount,
            averageTicket,
            DateTimeText.Iso(first),
            DateTimeText.Iso(last),
            daily,
            topProducts);
    }

    private IQueryable<ItemFact> PublishedItems(Guid? productId, DateTime fromBound, DateTime toBound)
    {
        var query =
            from si in _db.SaleItems.AsNoTracking()
            join s in _db.Sales.AsNoTracking() on si.SaleId equals s.Id
            join ap in _db.ArtifactPublications.AsNoTracking() on si.ParseAttemptId equals ap.ActiveParseAttemptId
            where s.OccurredAt >= fromBound && s.OccurredAt <= toBound
                  && (productId == null || si.ProductId == productId)
            select new ItemFact(si.SaleId, si.ProductId, si.Quantity, si.Total, s.OccurredAt);
        return query;
    }

    private sealed record ItemFact(Guid SaleId, Guid ProductId, decimal Quantity, decimal Total, DateTime? OccurredAt);
}

public sealed record DashboardResponse(
    string RevenueTotal,
    string QuantityTotal,
    long SalesCount,
    long ItemCount,
    string? AverageTicket,
    string? FirstMovementAt,
    string? LastMovementAt,
    IReadOnlyList<DailyPoint> Daily,
    IReadOnlyList<TopProduct> TopProducts);

public sealed record DailyPoint(string Date, string Quantity, string Revenue);

public sealed record TopProduct(
    Guid ProductId,
    string Name,
    string ExternalId,
    string Quantity,
    string Revenue);
